### ISSUE TITLE
Prevent too many set/extend* calls that run in parallel because not considering async nature

### ISSUE THREAD

**Issue by Apollon77** [ioBroker/ioBroker.js-controller#991]
We have many adapters that simply hammer "setObject/extendObject, setState, ..." calls at the same time or in loops which can end in "maximum concurrent callbacks" error or such and also bring low hardware systems to limits.

One idea is to build an internal "queue" for the affected set/extend calls for objects and states into the adapter base class that make sure that the calls are executed sequencially or max up to e.g. 10 in parallel.
Ideally we can change that mostly "Under the hood" so that adapter developers do not need to do anything.

Proposal:
* all relevant set/extend* calls on adapter methods are internally queued and the logic makes ure they are executed in the order they came in. Ideally we can configure a "maximum concurrent" number which allows to work on max x in parallel and this number is set to 10 by default
* manually setting this value to 0 when initializing the adapter class disables the queue and the adapte rbehaves exactly as now - this option is for special adapter specific needs and can be used by developers when tey know what they are doing.
* we add a new method "setQueueTrigger/Async" (better name preferred) that allows to get a Promise/callback when the specific point in the queue is reached where this trigger is inserted to e.g. do all setObjectNotExists at adapter start and after these are executed continue in code (is this working with the 10 in parallel approach?)

happy discussion

**Comment by AlCalzone** [comment]
I would do it under the hood using the old methods. If we introduce new ones, it will take years until all adapters use them. Implement a queue and thread pool in js controller, this way we can limit and control how many states are set in parallel. A pool of infinite size would equal the current behavior, 1 would be strictly sequential. In the end we could even configure this number to account for the system we are on.

I do not believe that this is a breaking change. During Zwave2 startup I set hundreds of states strictly sequentially and haven't noticed any bottlenecks.

**Comment by Apollon77** [comment]
I would also not see this as a breaking change... exactly as I thought - but with that "get a trigger at a certain point in the queue" method additionally as described

**Comment by AlCalzone** [comment]
I don't understand the use case for that trigger. Why would an adapter care for this? IMO the interesting part is when the setState is done.

**Comment by Apollon77** [comment]
Use case is as follows: When the adapter starts I want to create all objects and initialize the state values. The "subscribe(*) or such should not be done BEFORE all this is done, else I would need to add additional code to make sure everything is initialized or such. In fact it is kind of an "await "all queue stuff done till here"" and then execute the next part of the code.

I currently use small tooling in my adapters ... I add all objects and then I execute "processChanges" and this is executing the queue and  giving a callback after all this is done. I need exactly that. Mainly it is about code flow control

**Comment by AlCalzone** [comment]
You can already do that:

for (...) {
   await setStateAsync(...);
}

// stuff to do after initalizing

**Comment by Apollon77** [comment]
Yes I know, BUT we say here that we ewant to execute all "setState" Calls queued in the background to make it very easy for the devs to work like that ... and that they do not need to rewrite their adapters to async/await or such ... so ... think in that directly please.

**Comment by AlCalzone** [comment]
I think I need some pseudo code using the proposed method to understand what you mean.

For me it seems that two for loops like above should be enough for "do everything, then continue".
But maybe I'm missing the point here.

**Comment by AlCalzone** [comment]
I recently found something very interesting:
https://github.com/sindresorhus/p-queue

That module is basically a Promise task queue. Every call to the DB could just add a task to that queue and it would be executed as soon as there is capacity for it.

**Comment by Apollon77** [comment]
I like this idea ... we just need to move everything to promises then because  we currentlyconvert promises to callbacks and not other way ;-) Or?!

**Comment by AlCalzone** [comment]
We have callback-style APIs and need to wrap them in Promises (which we already do in many cases) to be able to use that.
